### ISSUE TITLE
Incorrect implementation of `_getPriceWithLowerChange()`

### ISSUE THREAD

**Issue by abhishekvispute** [raft-fi/contracts#364]
## Severity 
Medium

## Description 
`_getPriceWithLowerChange()` is used inside PriceFeed to return the price that is less changed between the two oracle's prices with respect to lastGoodPrice.

However, it's not being implemented correctly.
There are two different issues associated with it.

### Issue 1

```solidity
function _getPriceWithLowerChange(
        uint256 primaryOraclePrice,
        uint256 secondaryOraclePrice
    )
        internal
        view
        returns (uint256)
    {
       ## case 1
        if (primaryOraclePrice > lastGoodPrice && secondaryOraclePrice > lastGoodPrice) {
            return Math.min(primaryOraclePrice, secondaryOraclePrice);
        }
       ## case 2
        if (primaryOraclePrice < lastGoodPrice && secondaryOraclePrice < lastGoodPrice) {
            return Math.max(primaryOraclePrice, secondaryOraclePrice);
        }
        return lastGoodPrice;
    }
 ```
`_getPriceWithLowerChange` considers cases of `primaryOraclePrice` and `secondaryOraclePrice` both being greater or lesser than 
`lastGoodPrice`, but fail to consider the case where `lastGoodPrice` is in between them.
If it's in between, `lastGoodPrice` itself would be passed instead of the closest one to it as required.

![image](https://github.com/raft-fi/contracts/assets/46760063/dfd7079d-6687-4933-bdf3-d1454c06a630)

 `_getPriceWithLowerChange` is considered in volatile market conditions when both oracles are live, to find the closest price to the last good price.
**however, due to this bug, an old stale price would be considered even if both primary and secondary oracles are live.**

### Recommendation for Issue 1 
Consider taking the absolute difference between coordinates and selecting one with the closest distance.
```
DistancePrimary = Mod (lastGoodPrice - primaryOraclePrice)
DistanceSecondary = Mod (lastGoodPrice- secondaryOraclePrice)
If (DistancePrimary > DistanceSecondary) return secondaryOraclePrice
else return primaryOraclePrice
```

### Issue 2
If you resolve Issue 1, then this issue would automatically get resolved.
However, if you are still returning `lastGoodPrice` in-between case, consider resolving this issue.

The issue is a deviation in the case of `lastGoodPrice` return from `_getPriceWithLowerChange`, would always be considered `secondaryOracle.DEVIATION()`.
https://github.com/raft-fi/contracts/blob/aff9e2bdc9bf32a8512496eda07c498f10e30bb4/contracts/PriceFeed.sol#L86-L89
                
This contradicts, how deviation is considered in other places for lastGoodPrice.
In other places, as we can see max from secondary and primary deviation is considered if `lastGoodPrice` is returned.
https://github.com/raft-fi/contracts/blob/aff9e2bdc9bf32a8512496eda07c498f10e30bb4/contracts/PriceFeed.sol#L60
https://github.com/raft-fi/contracts/blob/aff9e2bdc9bf32a8512496eda07c498f10e30bb4/contracts/PriceFeed.sol#L72

### Recommendation for Issue 2
Consider correcting deviation derivation if you are keeping the return of `lastGoodPrice`  for the between case

**Comment by StefanVitor** [comment]
Hi @abhishekvispute 

Thanks for this submission and a very nice explanation, but this is not an issue. Our idea is next:
1. In case 1 and case 2, prices are on "same side" in regard to "lastGoodPrice", so in that case, we conclude that both prices are good in a volatile market, and we will pick price in a more conservative way (price with lower change)
2. In the case where one oracle price is less than "lastGoodPrice", and the second oracle price is greater than "lastGoodPrice", we definitely think that one of the oracles (or even both) are not reliable (and we cannot conclude which one is reliable), and in that case, we will return "lastGoodPrice" because that is a price that we trust.

On our side, we could improve comments in the code to explain better this, or maybe change the function name.

**Comment by abhishekvispute** [comment]
hey @StefanVitor, makes sense, in that case, Issue 2, should still hold, no?
I agree its not medium severity on the basis of issue 2 alone, but could be low no ?

**Comment by StefanVitor** [comment]
We confirm issue 2, and it is fixed in #366
